contrib/tap_sinks/udp_sink/source/udp_sink_impl.cc

```diff
@@ -158,6 +142,103 @@ void UdpTapSink::UdpTapSinkHandle::handleSocketStreamedTrace(
   }
 }
 
+size_t
+UdpTapSink::UdpTapSinkHandle::getEventBodysize(const envoy::data::tap::v3::SocketEvent& event,
+                                               envoy::config::tap::v3::OutputSink::Format format) {
+  size_t total_body_bytes = 0;
+  if (event.has_read()) {
+    if (format == envoy::config::tap::v3::OutputSink::JSON_BODY_AS_STRING) {
+      total_body_bytes = event.read().data().as_string().size();
+    } else {
+      total_body_bytes = event.read().data().as_bytes().size();
+    }
+  } else {
+    if (format == envoy::config::tap::v3::OutputSink::JSON_BODY_AS_STRING) {
+      total_body_bytes = event.write().data().as_string().size();
+    } else {
+      total_body_bytes = event.write().data().as_bytes().size();
+    }
+  }
+  return total_body_bytes;
+}
+
+void UdpTapSink::UdpTapSinkHandle::handleSocketStreamedTraceForMMultiEventsBigBody(
+    envoy::config::tap::v3::OutputSink::Format format,
+    const envoy::data::tap::v3::SocketEvent& event, uint64_t trace_id) {
+
+  // Create an new trace message with this event.
+  TapCommon::TraceWrapperPtr new_trace = std::make_unique<envoy::data::tap::v3::TraceWrapper>();
+  envoy::data::tap::v3::SocketStreamedTraceSegment& new_streamed_trace =
+      *new_trace->mutable_socket_streamed_trace_segment();
+
+  new_streamed_trace.set_trace_id(trace_id);
+  *new_streamed_trace.mutable_event() = event;
+
+  // Socket streamed trace with single event which the body size is bigger than 64K.
+  handleSocketStreamedTrace(std::move(new_trace), format);
+}
+
+void UdpTapSink::UdpTapSinkHandle::handleSocketStreamedTraceForMultiEvents(
+    TapCommon::TraceWrapperPtr&& trace, envoy::config::tap::v3::OutputSink::Format format) {
+
+  // Send directly because the total size of whole trace is less than 64K
+  size_t max_size_of_each_sub_data = static_cast<size_t>(parent_.getUdpMaxSendMsgDataSize(format));
+  size_t the_total_trace_size = static_cast<uint64_t>(trace->ByteSizeLong());
+  if (the_total_trace_size <= max_size_of_each_sub_data) {
+    doSubmitTrace(std::move(trace), format);
+    return;
+  }
+
+  envoy::data::tap::v3::SocketEvents* src_events =
+      trace->mutable_socket_streamed_trace_segment()->mutable_events();
+  auto* src_repeated_events = src_events->mutable_events();
+
+  // Handle the body size of single event is bigger than 64K.
+  for (int curr_event_index = src_repeated_events->size() - 1; curr_event_index >= 0;
+       --curr_event_index) {
+    const envoy::data::tap::v3::SocketEvent& curr_event =
+        src_repeated_events->Get(curr_event_index);
+    if (getEventBodysize(curr_event, format) > max_size_of_each_sub_data) {
+      // Slice and send the message
+      handleSocketStreamedTraceForMMultiEventsBigBody(
+          format, curr_event, trace->socket_streamed_trace_segment().trace_id());
+      // Event is handled and remove it.
+      src_repeated_events->DeleteSubrange(curr_event_index, 1);
```

Comment:
I think this may change the order of the data? Maybe it's unacceptable. And the logic actually is too complex.

I reviewed the previous code again, if we always need to merge or split the big event anyway, how about to use single even in the `PerSocketTapperImpl` directly. The low level read/event event number make no sense anyway because it's effected by too much factors.

/wait

Reply:
We should make the logic simpler. This is what the ideal logic in my mind:

a. check whether the input trace segment object size is less than the limited size. If yes, go to b, if not, go to c.
b. write the trace segment object out and return.
c. pick the event for start to the end, until the accumulated size is larger than the limited size. Then all previous events should be pack into a new trace segment object. And the left events could repeat above processing.